### ISSUE TITLE
SIGSEGV in EventPipeConfiguration::RegisterProvider on Linux

### ISSUE THREAD

**Issue by janvorli** [dotnet/coreclr#13689]
TL;DR: Threadpool threads are still running when the main thread is exiting. The global EventPipeConfiguration* EventPipe::s_pConfig is already cleared to NULL by EventPipe::Shutdown at that point and so when one of the secondary threads attempts to create an event provider, it crashes with SIGSEGV.

While trying to build the dotnet/cli repo on Ubuntu 14.04, I was hitting intermittent SIGSEGV that 
happened when running `/home/janvorli/git/cli/artifacts/linux-x64/stage2/dotnet` with ` remove -h -` argument. In happens in about 1 run out of ten ones and seems to happen when the app is exiting, since it dumps all the help info before it crashes.
Dump of the full stack with `clrstack -f` is below. Please note it is happening on a secondary thread.
```
OS Thread Id: 0x26ec (11)
        Child SP               IP Call Site
00007FFFEE9C0700 00007FFFF62EC4D0 libcoreclr.so!EventPipeConfiguration::RegisterProvider(EventPipeProvider&) + 48
00007FFFEE9C0740 00007FFFF62EBB1B libcoreclr.so!EventPipeInternal::CreateProvider(_GUID, void (*)(_GUID const*, unsigned int, unsigned char, unsigned long, unsigned long, void*, void*)) + 107
00007FFFEE9C07A0 00007FFF7CCD9C54
00007FFFEE9C07B0                  [InlinedCallFrame: 00007fffee9c07b0] System.Private.CoreLib.dll!System.Diagnostics.Tracing.EventPipeInternal.CreateProvider(System.Guid, EtwEnableCallback)
00007FFFEE9C07B0                  [InlinedCallFrame: 00007fffee9c07b0] System.Private.CoreLib.dll!System.Diagnostics.Tracing.EventPipeInternal.CreateProvider(System.Guid, EtwEnableCallback)
00007FFFEE9C07A0 00007FFF7CCD9C54 System.Private.CoreLib.dll!DomainNeutralILStubClass.IL_STUB_PInvoke(System.Guid, EtwEnableCallback) + 180
00007FFFEE9C0860 00007FFF7CD66A2E System.Private.CoreLib.dll!System.Diagnostics.Tracing.EventPipeEventProvider.System.Diagnostics.Tracing.IEventProvider.EventRegister(System.Guid ByRef, EtwEnableCallback, Void*, Int64 ByRef) + 46
00007FFFEE9C08A0 00007FFF7CCA3958 System.Private.CoreLib.dll!System.Diagnostics.Tracing.EventProvider.Register(System.Guid) + 120
00007FFFEE9C08E0 00007FFF7CC778A9 System.Private.CoreLib.dll!System.Diagnostics.Tracing.EventSource.Initialize(System.Guid, System.String, System.String[]) + 345
00007FFFEE9C0970 00007FFF7CC75CEA System.Private.CoreLib.dll!System.Diagnostics.Tracing.EventSource..ctor(System.Diagnostics.Tracing.EventSourceSettings, System.String[]) + 234
00007FFFEE9C09F0 00007FFF7E3A833D System.Diagnostics.DiagnosticSource.dll!System.Diagnostics.DiagnosticSourceEventSource..ctor() + 77
00007FFFEE9C0A00 00007FFF7E3A8627 System.Diagnostics.DiagnosticSource.dll!System.Diagnostics.DiagnosticSourceEventSource..cctor() + 23
00007FFFEE9C0A30 00007FFFF62C2B6B libcoreclr.so!DispatchCallDebuggerWrapper(CallDescrData*, ContextTransitionFrame*, int) + 203
00007FFFEE9C0AD0 00007FFFF62C2CED libcoreclr.so!DispatchCallSimple(unsigned long*, unsigned int, unsigned long, unsigned int) + 173
00007FFFEE9C0B60 00007FFFF6258A7D libcoreclr.so!MethodTable::RunClassInitEx(Object**) + 413
00007FFFEE9C0C60 00007FFFF62591A6 libcoreclr.so!MethodTable::DoRunClassInitThrowing() + 1190
00007FFFEE9C0D50                  [GCFrame: 00007fffee9c0d50]
00007FFFEE9C1650 00007FFFF648E535 libcoreclr.so!DynamicHelperFixup(TransitionBlock*, unsigned long*, unsigned int, Module*, CORCOMPILE_FIXUP_BLOB_KIND*, TypeHandle*, MethodDesc**, FieldDesc**) + 1509
00007FFFEE9C1C10 00007FFFF648F4E6 libcoreclr.so!DynamicHelperWorker + 326
00007FFFEE9C1CC8                  [DynamicHelperFrame: 00007fffee9c1cc8]
00007FFFEE9C1D30 00007FFFF640CAEF libcoreclr.so!DelayLoad_Helper + 119
00007FFFEE9C1E20 00007FFF7E3A76BA System.Diagnostics.DiagnosticSource.dll!System.Diagnostics.DiagnosticListener..ctor(System.String) + 138
00007FFFEE9C1E60 00007FFF7E22E7F1 System.Net.Http.dll!System.Net.Http.DiagnosticsHandler..cctor() + 33
00007FFFEE9C1E90 00007FFFF62C2B6B libcoreclr.so!DispatchCallDebuggerWrapper(CallDescrData*, ContextTransitionFrame*, int) + 203
00007FFFEE9C1F30 00007FFFF62C2CED libcoreclr.so!DispatchCallSimple(unsigned long*, unsigned int, unsigned long, unsigned int) + 173
00007FFFEE9C1FC0 00007FFFF6258A7D libcoreclr.so!MethodTable::RunClassInitEx(Object**) + 413
00007FFFEE9C20C0 00007FFFF62591A6 libcoreclr.so!MethodTable::DoRunClassInitThrowing() + 1190
00007FFFEE9C21B0                  [GCFrame: 00007fffee9c21b0]
00007FFFEE9C2AB0 00007FFFF648E535 libcoreclr.so!DynamicHelperFixup(TransitionBlock*, unsigned long*, unsigned int, Module*, CORCOMPILE_FIXUP_BLOB_KIND*, TypeHandle*, MethodDesc**, FieldDesc**) + 1509
00007FFFEE9C3070 00007FFFF648F4E6 libcoreclr.so!DynamicHelperWorker + 326
00007FFFEE9C3128                  [DynamicHelperFrame: 00007fffee9c3128]
00007FFFEE9C3190 00007FFFF640CAEF libcoreclr.so!DelayLoad_Helper + 119
00007FFFEE9C3280 00007FFF7E20E8A6 System.Net.Http.dll!System.Net.Http.HttpClientHandler.SendAsync(System.Net.Http.HttpRequestMessage, System.Threading.CancellationToken) + 38
00007FFFEE9C32C0 00007FFF7E2114F4 System.Net.Http.dll!System.Net.Http.HttpMessageInvoker.SendAsync(System.Net.Http.HttpRequestMessage, System.Threading.CancellationToken) + 132
00007FFFEE9C3300 00007FFF7E20B10F System.Net.Http.dll!System.Net.Http.HttpClient.SendAsync(System.Net.Http.HttpRequestMessage, System.Net.Http.HttpCompletionOption, System.Threading.CancellationToken) + 383
00007FFFEE9C3380 00007FFF7E20AF01 System.Net.Http.dll!System.Net.Http.HttpClient.SendAsync(System.Net.Http.HttpRequestMessage) + 65
00007FFFEE9C33B0 00007FFF7DEC209E Microsoft.ApplicationInsights.dll!Microsoft.ApplicationInsights.Channel.Transmission+<SendAsync>d__31.MoveNext() + 238
00007FFFEE9C3430 00007FFF7D152509 System.Private.CoreLib.dll!System.Runtime.CompilerServices.AsyncTaskMethodBuilder.Start[[Microsoft.ApplicationInsights.Channel.Transmission+<SendAsync>d__31, Microsoft.ApplicationInsights]](<SendAsync>d__31 ByRef) + 73
00007FFFEE9C3470 00007FFF7DEB0584 Microsoft.ApplicationInsights.dll!Microsoft.ApplicationInsights.Channel.Transmission.SendAsync() + 84
00007FFFEE9C34E0 00007FFF7DEC1E1E Microsoft.ApplicationInsights.dll!Microsoft.ApplicationInsights.Channel.InMemoryTransmitter+<Send>d__19.MoveNext() + 286
00007FFFEE9C3580 00007FFF7D13C309 System.Private.CoreLib.dll!System.Runtime.CompilerServices.AsyncTaskMethodBuilder.Start[[Microsoft.ApplicationInsights.Channel.InMemoryTransmitter+<Send>d__19, Microsoft.ApplicationInsights]](<Send>d__19 ByRef) + 73
00007FFFEE9C35C0 00007FFF7DEAFED8 Microsoft.ApplicationInsights.dll!Microsoft.ApplicationInsights.Channel.InMemoryTransmitter.Send(System.Collections.Generic.IEnumerable`1<Microsoft.ApplicationInsights.Channel.ITelemetry>) + 88
00007FFFEE9C3630 00007FFF7DEAFD4F Microsoft.ApplicationInsights.dll!Microsoft.ApplicationInsights.Channel.InMemoryTransmitter.DequeueAndSend() + 111
00007FFFEE9C36C0 00007FFF7DEAF8E1 Microsoft.ApplicationInsights.dll!Microsoft.ApplicationInsights.Channel.InMemoryChannel.Flush() + 17
00007FFFEE9C36D0 00007FFF7DEAB989 Microsoft.ApplicationInsights.dll!Microsoft.ApplicationInsights.TelemetryClient.Flush() + 25
00007FFFEE9C36E0 00007FFF7D1EC356 dotnet.dll!Microsoft.DotNet.Cli.Telemetry.Telemetry.TrackEventTask(System.String, System.Collections.Generic.IDictionary`2<System.String,System.String>, System.Collections.Generic.IDictionary`2<System.String,Double>) + 118
00007FFFEE9C3720 00007FFF7D1F3673 dotnet.dll!Microsoft.DotNet.Cli.Telemetry.Telemetry+<>c__DisplayClass14_0.<TrackEvent>b__0(System.Threading.Tasks.Task) + 35
00007FFFEE9C3730 00007FFF7CC07FA1 System.Private.CoreLib.dll!System.Threading.ExecutionContext.Run(System.Threading.ExecutionContext, System.Threading.ContextCallback, System.Object) + 113
00007FFFEE9C3780 00007FFF7CC3FBC4 System.Private.CoreLib.dll!System.Threading.Tasks.Task.ExecuteWithThreadLocal(System.Threading.Tasks.Task ByRef) + 372
00007FFFEE9C3810 00007FFF7CC8284E System.Private.CoreLib.dll!System.Threading.ThreadPoolWorkQueue.Dispatch() + 414
00007FFFEE9C3880 00007FFFF640C86B libcoreclr.so!CallDescrWorkerInternal + 124
00007FFFEE9C38A0 00007FFFF62C3590 libcoreclr.so!MethodDescCallSite::CallTargetWorker(unsigned long const*, unsigned long*, int) + 1264
00007FFFEE9C3A90 00007FFFF64AF2D6 libcoreclr.so!QueueUserWorkItemManagedCallback(void*) + 198
00007FFFEE9C3BB0 00007FFFF6294BA2 libcoreclr.so!ManagedThreadBase_DispatchOuter(ManagedThreadCallState*) + 402
00007FFFEE9C3C00                  [DebuggerU2MCatchHandlerFrame: 00007fffee9c3c00]
00007FFFEE9C3CD0 00007FFFF6295386 libcoreclr.so!ManagedThreadBase::ThreadPool(ADID, void (*)(void*), void*) + 70
00007FFFEE9C3D20 00007FFFF6492677 libcoreclr.so!ManagedPerAppDomainTPCount::DispatchWorkItem(bool*, bool*) + 295
00007FFFEE9C3E10 00007FFFF62B382E libcoreclr.so!ThreadpoolMgr::WorkerThreadStart(void*) + 1134
00007FFFEE9C3EC0 00007FFFF662DE42 libcoreclr.so!CorUnix::CPalThread::ThreadEntry(void*) + 306
00007FFFEE9C3EE0 00007FFFF79C0184 libpthread.so.0!start_thread + 196 at /build/eglibc-SvCtMH/eglibc-2.19/nptl/pthread_create.c:312
00007FFFEE9C3F80 00007FFFF6ECCFFD libc.so.6!__clone + 109 at /build/eglibc-SvCtMH/eglibc-2.19/misc/../sysdeps/unix/sysv/linux/x86_64/clone.S:113
```

Disass at the failing frame looks like this:
```asm
disass -F intel
libcoreclr.so`EventPipeConfiguration::RegisterProvider:
    0x7ffff62ec4a0 <+0>:   push   rbp
    0x7ffff62ec4a1 <+1>:   mov    rbp, rsp
    0x7ffff62ec4a4 <+4>:   push   r15
    0x7ffff62ec4a6 <+6>:   push   r14
    0x7ffff62ec4a8 <+8>:   push   r13
    0x7ffff62ec4aa <+10>:  push   r12
    0x7ffff62ec4ac <+12>:  push   rbx
    0x7ffff62ec4ad <+13>:  push   rax
    0x7ffff62ec4ae <+14>:  mov    r15, rsi
    0x7ffff62ec4b1 <+17>:  mov    r12, rdi
    0x7ffff62ec4b4 <+20>:  call   0x7ffff62eb410            ; EventPipe::GetLock()
    0x7ffff62ec4b9 <+25>:  mov    qword ptr [rbp - 0x30], rax
    0x7ffff62ec4bd <+29>:  mov    rdi, rax
    0x7ffff62ec4c0 <+32>:  call   0x7ffff6211800            ; CrstBase::Enter()
    0x7ffff62ec4c5 <+37>:  mov    rdi, r15
    0x7ffff62ec4c8 <+40>:  call   0x7ffff62edd40            ; EventPipeProvider::GetProviderID() const
    0x7ffff62ec4cd <+45>:  mov    r13, rax
->  0x7ffff62ec4d0 <+48>:  mov    rax, qword ptr [r12 + 0x18]
```
The R12 register contains zero at this point. As you can see, since it is a callee saved register, it should hold the value of the RDI argument register, that means the first argument of the function, that means the "this" pointer. 

Looking at the frame 1, I can see that this function was actually tail called by the following function:
```asm
libcoreclr.so`EventPipeProvider::EventPipeProvider:
    0x7ffff62edc20 <+0>:   push   rbp
    0x7ffff62edc21 <+1>:   mov    rbp, rsp
    0x7ffff62edc24 <+4>:   push   r15
    0x7ffff62edc26 <+6>:   push   r14
    0x7ffff62edc28 <+8>:   push   rbx
    0x7ffff62edc29 <+9>:   push   rax
    0x7ffff62edc2a <+10>:  mov    r14, rcx
    0x7ffff62edc2d <+13>:  mov    r15, rdx
    0x7ffff62edc30 <+16>:  mov    rbx, rdi
    0x7ffff62edc33 <+19>:  movups xmm0, xmmword ptr [rsi]
    0x7ffff62edc36 <+22>:  movups xmmword ptr [rbx], xmm0
    0x7ffff62edc39 <+25>:  mov    byte ptr [rbx + 0x10], 0x0
    0x7ffff62edc3d <+29>:  mov    qword ptr [rbx + 0x18], 0x0
    0x7ffff62edc45 <+37>:  mov    dword ptr [rbx + 0x20], 0x1
    0x7ffff62edc4c <+44>:  mov    edi, 0x18
    0x7ffff62edc51 <+49>:  call   0x7ffff61e68a0            ; operator new(unsigned long)
    0x7ffff62edc56 <+54>:  mov    qword ptr [rax], 0x0
    0x7ffff62edc5d <+61>:  mov    qword ptr [rax + 0x8], rax
    0x7ffff62edc61 <+65>:  mov    qword ptr [rax + 0x10], rax
    0x7ffff62edc65 <+69>:  mov    qword ptr [rbx + 0x28], rax
    0x7ffff62edc69 <+73>:  mov    qword ptr [rbx + 0x30], r15
    0x7ffff62edc6d <+77>:  mov    qword ptr [rbx + 0x38], r14
    0x7ffff62edc71 <+81>:  call   0x7ffff62eb770            ; EventPipe::GetConfiguration()
    0x7ffff62edc76 <+86>:  mov    qword ptr [rbx + 0x40], rax
    0x7ffff62edc7a <+90>:  mov    rdi, rax
    0x7ffff62edc7d <+93>:  mov    rsi, rbx
    0x7ffff62edc80 <+96>:  add    rsp, 0x8
    0x7ffff62edc84 <+100>: pop    rbx
    0x7ffff62edc85 <+101>: pop    r14
    0x7ffff62edc87 <+103>: pop    r15
    0x7ffff62edc89 <+105>: pop    rbp
    0x7ffff62edc8a <+106>: jmp    0x7ffff62ec4a0            ; EventPipeConfiguration::RegisterProvider(EventPipeProvider&)
```
As you can see, the RDI register contains the result of the EventPipe::GetConfiguration() function call, so that one must have returned NULL.
The function is quite simple:
```asm
libcoreclr.so`EventPipe::GetConfiguration:
    0x7ffff62eb770 <+0>:  mov    rax, qword ptr [rip + 0x507719] ;  + 6096
    0x7ffff62eb777 <+7>:  mov    rax, qword ptr [rax]
    0x7ffff62eb77a <+10>: ret
```
And really, dumping the value that it would return shows it is NULL:
```
(lldb) x/gx 0x7ffff62eb777+0x507719
0x7ffff67f2e90: 0x00007ffff6831a10
(lldb) x/gx 0x00007ffff6831a10
0x7ffff6831a10: 0x0000000000000000
```
The interesting thing is the stack of the main thread:
```
* thread #1: tid = 9952, 0x00007ffff7de40c8 ld-linux-x86-64.so.2`??? + 56, name = 'dotnet'
  * frame #0: 0x00007ffff7de40c8 ld-linux-x86-64.so.2`??? + 56
    frame #1: 0x00007ffff7de4b31 ld-linux-x86-64.so.2`??? + 305
    frame #2: 0x00007ffff7de96f7 ld-linux-x86-64.so.2`??? + 247
    frame #3: 0x00007ffff7df06a5 ld-linux-x86-64.so.2`??? + 53
    frame #4: 0x00007fff4c468563 libcom_err.so.2
    frame #5: 0x00007ffff7dea8da ld-linux-x86-64.so.2`??? + 474
    frame #6: 0x00007ffff6e0b1a9 libc.so.6`__run_exit_handlers(status=0, listp=0x00007ffff71916c8, run_list_atexit=<unavailable>) + 217 at exit.c:82
    frame #7: 0x00007ffff6e0b1f5 libc.so.6`__GI_exit(status=<unavailable>) + 21 at exit.c:104
    frame #8: 0x00007ffff6df0f4c libc.so.6`__libc_start_main(main=<unavailable>, argc=<unavailable>, argv=<unavailable>, init=<unavailable>, fini=<unavailable>, rtld_fini=<unavailable>, stack_end=0x00007fffffffe428) + 252 at libc-start.c:321
    frame #9: 0x000000000040a335 dotnet`_start + 41
```
As you can see, the thread has already exited the "main" function and is running exit handlers.

**Comment by janvorli** [comment]
CC: @weshaggard, @danmosemsft, @eerhardt, @ericstj - I think one of you guys have mentioned in our meeting on Monday that you have seen dotnet tool crashing randomly from time to time. I think this might be the actual issue.

**Comment by danmosemsft** [comment]
Yes I've seen it crashing during build. Good!

**Comment by brianrob** [comment]
@janvorli thanks for finding this.  I'll get working on a fix.  Have you only seen this repro when building the CLI?  I want to make sure that I can repro this locally.

**Comment by janvorli** [comment]
@brianrob yes, I've found it when building the freshly synced cli master branch on my Ubuntu 14.04. The command to build the cli stuff was `./build.sh --skip-prereqs --configuration Release --targets Default`.
Some of the tests that the build runs were failing due to this issue. You can run `cli/artifacts/linux-x64/stage2/dotnet with remove -h -` after the build completes to repro it. As I've said it reproed in about 1 out of 10 runs for me. Not sure if it matters, but my Linux box is a 8 core Xeon E5530 with 24GB of RAM
Btw, the libcoreclr.so that it runs with was built from coreclr repo at commit https://github.com/dotnet/coreclr/commit/dc2ffcdff2fb19e4b252cec3bc6672400a375c1e

**Comment by nguerrera** [comment]
We have seen SIGSEGV in build on Jenkins, there's some old threads on that. I've never seen it locally (though I don't build that much on Linux) and even on Jenkins, it's been a long while since I've seen it

**Comment by brianrob** [comment]
Thanks @janvorli.  I'm going to try and repro this.

**Comment by brianrob** [comment]
@janvorli, just confirming that you were able to repro this via "./dotnet -h" on a CLI release build.  I am doing this on an 8 core machine with 16 GB of RAM and can't repro it so far.  It uses the same CoreCLR that you referenced above.  Just want to make sure that I have the repro setup so far.

**Comment by janvorli** [comment]
not just dotnet -h, it was `dotnet remove -h -` which is help on the "remove" command.

**Comment by brianrob** [comment]
Gotcha.  Thanks.  That doesn't seem to repro it either.   Do you have a machine that I can use to repro?  I'm less concerned about seeing the failure since I see your analysis, but I want to make sure that any fix I write is sufficient.

**Comment by janvorli** [comment]
Unfortunately my machine is not on corpnet, so you cannot access it. I assume you have tried to run it in a loop and it never reproed?
This is a command I used (current directory being the root of the cli repo):
```
for i in `seq 1 10000`; do ./artifacts/linux-x64/stage2/dotnet remove -h -; if [ $? -eq 139 ]; then break; fi; done
```
It takes only a few iterations to exit due to the SIGSEGV

**Comment by brianrob** [comment]
I have tried to run it in a loop.  I'll let it run for a while longer and see if anything repros.

**Comment by janvorli** [comment]
Quick question, what do you get from the following command (I wonder if you have the same version of the shared framework)
`ls ./artifacts/linux-x64/stage2/shared/Microsoft.NETCore.App`

**Comment by janvorli** [comment]
@brianrob I am still hitting this issue  when building / testing cli repo, now also in CentOS 6 docker container running on the Ubuntu 14.04 where I was hitting it originally. Have you made any progress fixing it?

**Comment by brianrob** [comment]
@janvorli, sorry for the delay.  I've just posted a PR for this and tagged you on it.
